extensions-core/lookups-cached-single/src/main/java/io/druid/server/lookup/PollingLookup.java

```diff
@@ -0,0 +1,242 @@
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied. See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ * /
+ */
+
+package io.druid.server.lookup;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningScheduledExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.metamx.common.ISE;
+import com.metamx.common.logger.Logger;
+import io.druid.concurrent.Execs;
+import io.druid.query.lookup.LookupExtractor;
+import io.druid.server.lookup.cache.polling.OnHeapPollingCache;
+import io.druid.server.lookup.cache.polling.PollingCache;
+import io.druid.server.lookup.cache.polling.PollingCacheFactory;
+
+import javax.validation.constraints.NotNull;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class PollingLookup extends LookupExtractor
+{
+  private static final Logger LOGGER = new Logger(PollingLookup.class);
+
+  private final long pollPeriodMs;
+
+  private final DataFetcher dataFetcher;
+
+  private final PollingCacheFactory cacheFactory;
+
+  private final AtomicReference<CacheRefKeeper> refOfCacheKeeper = new AtomicReference<>();
+
+
+  private final ListeningScheduledExecutorService scheduledExecutorService;
+
+  private final AtomicBoolean isOpen = new AtomicBoolean(false);
+
+  private final ListenableFuture<?> pollFuture;
+
+  private final String id =  Integer.toHexString(System.identityHashCode(this));
+
+  public PollingLookup(
+      long pollPeriodMs,
+      DataFetcher dataFetcher,
+      PollingCacheFactory cacheFactory
+  )
+  {
+    this.pollPeriodMs = pollPeriodMs;
+    this.dataFetcher = Preconditions.checkNotNull(dataFetcher);
+    this.cacheFactory = cacheFactory == null ? new OnHeapPollingCache.OnHeapPollingCacheProvider() : cacheFactory;
+    refOfCacheKeeper.set(new CacheRefKeeper(this.cacheFactory.makeOf(dataFetcher.fetchAll())));
+    if (pollPeriodMs > 0) {
+      scheduledExecutorService = MoreExecutors.listeningDecorator(Executors.newSingleThreadScheduledExecutor(
+          Execs.makeThreadFactory("PollingLookup-" + id, Thread.MIN_PRIORITY)
+      ));
+      pollFuture = scheduledExecutorService.scheduleAtFixedRate(
+          pollAndSwap(),
+          pollPeriodMs,
+          pollPeriodMs,
+          TimeUnit.MILLISECONDS
+      );
+    } else {
+      scheduledExecutorService = null;
+      pollFuture = null;
+    }
+    this.isOpen.set(true);
+  }
+
+
+  public void close()
```

Comment:
All of this and the stuff in the constructor feels like it should be in the factory instead of here. Is there a reason it shouldn't be in the factory, which has lifecycle assumptions already built into it?